### PR TITLE
TypeScript definitions

### PR DESCRIPTION
**Pull request by darky**

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  [key: string]: INothing;
+}
+
+interface IFunctionProps {
+  length: 0;
+  name: string;
+  prototype: object;
+  toLocaleString: "";
+  toString: "";
+  valueOf(): false;
```
Comment (slmgc) [inline index.d.ts]: Should `valueOf` have parentheses?

Comment (darky) [reply index.d.ts]: Yep, I forget parentheses on `toString` and `toLocaleString`
Already added
Thanks for notice, good catch!

**Comment by slmgc** [comment]
Looks good to me, thanks!

**Comment by slmgc** [comment]
I am not a typescript user, but after some testing I found a few issues with these definitions:

1. `Nothing.length = 2 // Cannot assign to 'length' because it is a constant or a read-only property.` - it should be assignable though, this is related to all props 
1. `Nothing.anyProperty = {} // Type '{}' is not assignable to type 'INothing'.` - any property should allow assignment
1. `(p1: any, p2: any, p3: any, ...): INothing` - all these declarations can be replaced by `(...args: any[]): INothing`
1. `Nothing().call() // Expected at least 1 arguments, but got 0.` - should allow any number of arguments

As a summary, I think it's better to replace all TS declarations with something like this:
```
export declare const Nothing: any
export declare function toBool (p: any): boolean
export declare function isNothing (p: any): boolean
export declare function isSomething (p: any): boolean
export declare function serialize (p: object): string
export declare function deserialize (p: string): object
```

Otherwise, it's too restrictive and has a lot of edge cases.

**Comment by darky** [comment]
1-4 cases good catches and should be fixed.

```
export declare const Nothing: any
```

I think it's bad idea. `any` so weak. It's almost equal to case, where no definitions totally.
Better to pick Nothing as particular nominal type and operate with it.

```
export declare function toBool (p: any): boolean
export declare function isNothing (p: any): boolean
export declare function isSomething (p: any): boolean
```

This has the meaning for older TypeScript versions without conditional types. Now, with conditional types and Nothing as particular nominal type, you can achieve more robust behavior from compiler.

**Comment by darky** [comment]
Ok, I investigate 1-4 problems:

1. It's native *lib.d.ts* TypeScript definition and can't be fixed:

```typescript
interface Function {
    readonly length: number;
}
```

Interface `INothing` extended from `Function` interface, because it callable.

2. Seems TypeScript not support co/contr-variance of interface properties, like FlowType. It assume "mono-type" on input/output. In our case it `INothing`. This can be easy fixed via `Nothing.anyProp = {} as any`. So popular practice.

3. It's understand, ok.

4. This related to (1). Also *lib.d.ts*, `thisArg` required.

```typescript
interface Function {
    call(this: Function, thisArg: any, ...argArray: any[]): any;
}
```

My opinion, a little embarrassing (2) case. But it's not problem for extended TypeScript user.
If it discouraged you, feel free to use own suggestion with `any`

**Comment by slmgc** [comment]
Thanks for the explanation! I guess I'll do the change as the current definitions have too strict restrictions in some cases, IMHO.

**Pull request merged**